homeassistant/components/hassmpris/media_player.py

```diff
@@ -0,0 +1,674 @@
+"""Support for interfacing with the HASS MPRIS agent."""
+from __future__ import annotations
+
+import asyncio
+import json
+import re
+from typing import Any, cast
+
+from hassmpris.proto import mpris_pb2
+import hassmpris_client
+import voluptuous as vol
+
+from homeassistant.components.media_player import (
+    MediaPlayerDeviceClass,
+    MediaPlayerEntity,
+    MediaPlayerEntityFeature,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import (
+    STATE_IDLE,
+    STATE_OFF,
+    STATE_PAUSED,
+    STATE_PLAYING,
+    STATE_UNKNOWN,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import config_validation as cv, entity_registry as er
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+import homeassistant.util.dt as dt_util
+
+from .const import (
+    ATTR_PLAYBACK_RATE,
+    DOMAIN,
+    ENTRY_CLIENT,
+    ENTRY_MEDIA_PLAYER_ENTITY_MANAGER,
+    ENTRY_UNLOADERS,
+    LOGGER as _LOGGER,
+)
+
+PLATFORM = "media_player"
+
+DISCOVERY_SCHEMA = vol.Schema(
+    {
+        vol.Required("player_id"): cv.string,
+    }
+)
+
+SUPPORTED_MINIMAL = (
+    MediaPlayerEntityFeature.PAUSE
+    | MediaPlayerEntityFeature.PLAY
+    | MediaPlayerEntityFeature.STOP
+)
+SUPPORTED_TURN_OFF = MediaPlayerEntityFeature.TURN_OFF
+SUPPORTED_TURN_ON = MediaPlayerEntityFeature.TURN_ON
+
+# Enable to remove clones (second / third, nth instances)
+# of any player when the player exits.  Not enabled by
+# default because events in the logbook disappear when
+# the player goes away.
+# Clones are always removed upon start or reload of the
+# integration, since there is no value in keeping them
+# around.
+REMOVE_CLONES_WHILE_RUNNING = False
+
+
+def _feat2bitfield(bitfield: int, obj: object) -> str:
+    fields = []
+    for feat in dir(obj):
+        val = getattr(obj, feat)
+        if not isinstance(val, int):
+            continue
+        if bitfield & val:
+            fields.append(feat)
+    return ",".join(fields)
+
```

Comment:
This looks like it might be able to be moved to the library

Reply:
Absolutely.  Done.  Hang on for the push.